vernac/declare.ml

```diff
@@ -192,35 +192,35 @@ let { Goptions.get = private_poly_univs } =
     ~value:true
     ()
 
-let universes_of_body_type body typ =
+let universes_of_body_type ~used_univs body typ =
   let used_univs_body = Vars.universes_of_constr body in
-  let used_univs_typ = Option.cata Vars.universes_of_constr Univ.Level.Set.empty typ in
+  let used_univs_typ = Univ.Level.Set.union used_univs (Option.cata Vars.universes_of_constr Univ.Level.Set.empty typ) in
   let used_univs = Univ.Level.Set.union used_univs_body used_univs_typ in
   used_univs_typ, used_univs
 
 let make_univs_deferred_private_mono ~initial_euctx ?feedback_id ~uctx ~udecl body typ =
-  let _, used_univs = universes_of_body_type body typ in
+  let _, used_univs = universes_of_body_type ~used_univs:Univ.Level.Set.empty body typ in
   let uctx = UState.constrain_variables (fst (UState.context_set initial_euctx)) uctx in
   (* For vi2vo compilation proofs are computed now but we need to
      complement the univ constraints of the typ with the ones of
      the body.  So we keep the two sets distinct. *)
   let uctx_body = UState.restrict uctx used_univs in
   UState.check_mono_univ_decl uctx_body udecl
 
-let make_univs_immediate_private_mono ~initial_euctx ~uctx ~udecl ~eff body typ =
+let make_univs_immediate_private_mono ~initial_euctx ~uctx ~udecl ~eff ~used_univs body typ =
   let utyp = UState.univ_entry ~poly:false initial_euctx in
+  let _, used_univs = universes_of_body_type ~used_univs body typ in
   let ubody =
-    let _, used_univs = universes_of_body_type body typ in
     let uctx = UState.constrain_variables (fst (UState.context_set initial_euctx)) uctx in
     (* For vi2vo compilation proofs are computed now but we need to
        complement the univ constraints of the typ with the ones of
        the body.  So we keep the two sets distinct. *)
     let uctx_body = UState.restrict uctx used_univs in
     UState.check_mono_univ_decl uctx_body udecl in
-  initial_euctx, utyp, Default { body = (body, eff); opaque = Opaque ubody }
+  initial_euctx, utyp, used_univs, Default { body = (body, eff); opaque = Opaque ubody }
```

Comment:
The returned tuple is getting big, maybe define a record type for it?

Reply:
I think I'm fine as it is (especially since the tupling is immediately destructed by the caller) but I can use a tuple if your question leans on the side of an assertion.